### DISCUSSION TITLE
Extend tasks with stuff that needs additional scripts?

### DISCUSSION THREAD

**Discussion by ModernRonin** [sharpliner/sharpliner#170] [Q&A] [answered]
Hi,

Background: the docker task doesn't allow to run containers. You are supposed to use a container-job for that. However, container jobs don't let you run Docker inside of Docker.

For my use-case, I need docker-in-docker, though. So I solved this with a custom task that calls docker via a powershell script and passes the "magic" command-line arguments enabling docker-in-docker. 

The question is would you see stuff like this being part of your library, too?

I see pros and cons:
PRO: 
* other people will have this scenario, too, and having this in Sharpliner will save them a few hours of research and guess-and-try.
CON:
* it's something that's not included in DevOps, so might be out-of-scope
* if docker ever changes their command-line arguments, said script will have to be adapted, so all of a sudden maintaining Sharpliner means staying in sync with *two* sources of change instead of one (Devops itself).

**Comment by premun** [comment]
Hi,

Thanks for the idea. My current idea is that I would like to keep the main Sharpliner only follow the basic YAML schema.
On the other hand, I am already improving things (such as the DotNet task) because it brings a lot of value to have nice APIs.

I was wondering if we could maybe make an Extensions project/nuget which could become a place for these things. However, it is unclear to me, how much can be external and how much of these things would require access to the internal Sharpliner pieces. I am also thinking about how opinionated things should be. Also there is no community around Sharpliner really, so maybe a marketplace and extensions would be too much at this point.

For the issue of going out of sync - we can always lock the version of the tasks we're creating and thus staying current.

And to your specific case - I am not sure whether this is not just a too specific scenario?

**Reply by ModernRonin** [reply]
I think the idea of extensions is very good. And that doesn't have to mean anything more than making it easy for folk to author packages with additional tasks, so no marketplace etc. 

as for community, I think Sharpliner has enormous potential. Once I'm not completely overloaded, I'll write an article about it on my blog, a few people read it regularly - really very few, like ~10 regulars plus a lot of random people who mostly find it because of projectrenamer), but even so this might attract more folks. I think this project deserves to live a long and happy life :-)

**Comment by premun** [comment, marked as answer]
Closing this discussion as we can just keep putting more extensions directly inside `Sharpliner` or introduce `Sharpliner.Extensions` (or similar) nuget.
